Source/WebCore/rendering/RenderBox.cpp

```diff
@@ -3647,8 +3647,14 @@ void RenderBox::computeBlockDirectionMargins(const RenderBlock& containingBlock,
     // the containing block (8.3)
     LayoutUnit cw = containingBlockLogicalWidthForContent();
     const RenderStyle& containingBlockStyle = containingBlock.style();
-    marginBefore = minimumValueForLength(style().marginBeforeUsing(&containingBlockStyle), cw);
-    marginAfter = minimumValueForLength(style().marginAfterUsing(&containingBlockStyle), cw);
+    if (containingBlock.shouldTrimChildMargin(MarginTrimType::BlockStart, *this))
+        marginBefore = 0_lu;
+    else
+        marginBefore = minimumValueForLength(style().marginBeforeUsing(&containingBlockStyle), cw);
+    if (containingBlock.shouldTrimChildMargin(MarginTrimType::BlockEnd, *this))
+        marginAfter = 0_lu;
+    else
+        marginAfter = minimumValueForLength(style().marginAfterUsing(&containingBlockStyle), cw);
```

Comment:
I wonder if this pattern is common enough to warrant a helper function.

Reply:
Yes I think so. I ended up needing something similar for the Flexbox implementation so I will add one

Reply:
I added a new helper method, `constrainBlockMarginInAvailableSpaceOrTrim` let me know what you think. We may be able to move the computation of the available size to inside the method itself, but I took a quick look at my flexbox patch and it looks like there was a case where it was overridden to a different value.